### PR TITLE
feat: add `@AllowNull`, `@AutoIncrement`, `@Comment`, `@Default`, `@PrimaryKey`, and validation decorators

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

I reworked our internal decorator utils to reduce duplication of code

This PR adds the following decorators:

```
AllowNull
AutoIncrement
ColumnName
Comment
Default
NotNull
PrimaryKey

ValidateAttribute
ModelValidator
Contains
Equals
Is
IsAfter
IsAlpha
IsAlphanumeric
IsArray
IsBefore
IsCreditCard
IsDate
IsDecimal
IsEmail
IsFloat
IsIP
IsIPv4
IsIPv6
IsIn
IsInt
IsLowercase
IsNumeric
IsUUID
IsUppercase
IsUrl
Len
Max
Min
Not
NotContains
NotEmpty
NotIn
```

A few differences with sequelize-typescript:

- `IsNull` has been removed
- `NotNull` configures the "allowNull" option instead of `validation.notNull`. I've marked `validation.notNull` and `validation.isNull` for future removal, as it makes no sense to have three ways to validate whether something is or isn't null
- validation and hook decorators can be used on symbol properties ✨
- `Validate` has been renamed to `ValidateAttribute` 
- `Validator` has been renamed to `ModelValidator`
  - Static model validators don't receive the instance to validate through "this", but as the first parameter.
  - Instance model validators remain the same: they receive it through "this"
- I've added `ColumnName` as a more user friendly name than `field` (until we rename `field` to `columnName`)

### CODE REVIEW

File: test/unit/esm-named-exports.test.js [hunk 1]
```diff
@@ -9,104 +9,122 @@ const expect = chai.expect;
...
+      for (const key of cjsKeys) {
+        if (ignoredCjsKeys.includes(key)) {
+          continue;
+        }
 
-    expect(missingEsmKeys.length).to.eq(0,
-      `esm entry point is missing exports: ${missingEsmKeys.map(v => JSON.stringify(v)).join(', ')}.
-Either add these exports to "index.mjs" (and "types/index.d.ts"), or mark them as ignored in "esm-named-exports.test.js"
-      `);
+        if (!esmKeys.includes(key)) {
```
Comment (evanrittenhouse) [inline test/unit/esm-named-exports.test.js:96]: We're looking for missing ESM keys here - do we want to check for any missing CJS ones for backwards compatibility (e.g. mistaken `export default` not being caught in code reviews)? It's not the express purpose of this file, but may be helpful. Happy to disregard if there's not a good reason for it though

File: test/unit/decorators/attribute.test.ts [hunk 2]
```diff
@@ -206,3 +206,204 @@ describe(`@Attribute legacy decorator`, () => {
...
+
+    sequelize.addModels([User]);
+
+    expect(User.getAttributes().name.allowNull).to.equal(false);
+  });
+
+  it('accepts a boolean', () => {
+    class User extends Model<InferAttributes<User>> {
+      @Attribute(DataTypes.STRING)
+      @NotNull(false)
```
Comment (evanrittenhouse) [inline test/unit/decorators/attribute.test.ts:252]: nit: May be worth only accepting booleans on `AllowNull` - IMO `NotNull` explicitly means "this cannot be null", whereas allow null means "Do we allow this to be null? Default to yes"

File: src/decorators/legacy/attribute-utils.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  decoratorName: string,
+  defaultValue: T | typeof DECORATOR_NO_DEFAULT,
+  callback: (
+    option: T,
+    target: Object,
+    propertyName: string | symbol,
+    propertyDescriptor: PropertyDescriptor | undefined,
+  ) => Partial<ModelAttributeColumnOptions>,
+): OptionalParameterizedPropertyDecorator<T> {
+  return createOptionallyParameterizedPropertyDecorator(
```
Comment (evanrittenhouse) [inline src/decorators/legacy/attribute-utils.ts:50]: Are we specifying attributes as model attributes and properties as POJO properties?

File: src/decorators/legacy/decorator-utils.ts [hunk 4]
```diff
@@ -4,41 +4,121 @@ export type PropertyOrGetterDescriptor = (
...
     target: Object,
     propertyName: string | symbol,
     propertyDescriptor: PropertyDescriptor | undefined,
   ) => void,
-): ParameterizedPropertyDecorator<T> {
-  return function decorator(...args: [options: T] | Parameters<PropertyOrGetterDescriptor>) {
-    if (args.length === 1) {
+): OptionalParameterizedPropertyDecorator<T> {
+  return function decorator(...args: [] | [options: T] | Parameters<PropertyOrGetterDescriptor>) {
+    // note: cannot use <= 1, because TypeScript uses this to infer the type of "args".
```
Comment (evanrittenhouse) [inline src/decorators/legacy/decorator-utils.ts:59]: Purely for my own curiosity, why's this?

File: src/decorators/legacy/decorator-utils.ts [hunk 5]
```diff
@@ -4,41 +4,121 @@ export type PropertyOrGetterDescriptor = (
...
+    if (args.length === 0 || args.length === 1) {
```
Comment (evanrittenhouse) [inline src/decorators/legacy/decorator-utils.ts:60]: These signatures can get a bit unwieldy, we may want to create some sort of DecoratorOptionsDTO or something to clear up a lot of these methods

**Review by evanrittenhouse — COMMENTED** [review]
Minor comments, but wow this is some excellent code.

File: src/instance-validator.js [hunk 6]
```diff
@@ -2,6 +2,8 @@
 
 import { AbstractDataType } from './dialects/abstract/data-types';
 import { validateDataType } from './dialects/abstract/data-types-utils';
+import { combinedIterator } from './utils/iterators';
```
Comment (WikiRik) [inline src/instance-validator.js:5]: This is unused

File: test/unit/esm-named-exports.test.js [hunk 7]
```diff
@@ -9,104 +9,122 @@ const expect = chai.expect;
 
 const nodeMajorVersion = Number(process.version.match(/(?<=^v)\d+/));
```
Comment (WikiRik) [inline test/unit/esm-named-exports.test.js:10]: This is unused

**Review by WikiRik — COMMENTED** [review]
Two unused imports but no further new comments from my side. Looking really good and also some nice improvements on the hook decorators!

File: src/decorators/legacy/decorator-utils.ts (hunk 5 above)
Comment (ephys) [reply src/decorators/legacy/decorator-utils.ts:60]: Decorators are so annoying to implement, I'm so annoyed they didn't take the years of feedback from the legacy decorators into account and still went with the same super annoying design of not auto-inserting parentheses (https://github.com/tc39/proposal-decorators/issues/81#issuecomment-713167505)

Anyway, what do you have in mind to clean this up?

File: src/decorators/legacy/attribute-utils.ts (hunk 3 above)
Comment (ephys) [reply src/decorators/legacy/attribute-utils.ts:50]: yep, model attributes means the properties of a model that are mapped to something to the DB, and property is anything that's in a class

File: test/unit/decorators/attribute.test.ts (hunk 2 above)
Comment (ephys) [reply test/unit/decorators/attribute.test.ts:252]: I don't expect it to be used but I did it for symmetry with the other boolean decorators, such as `IsFloat`, `IsDecimal` etc

We could make a `createUnparameterizedAttributeValidationDecorator` and `createUnparameterizedAttributeOptionsDecorator`, but I figured we already had too many decorator-making functions

I think we just won't document it :)

File: test/unit/esm-named-exports.test.js (hunk 1 above)
Comment (ephys) [reply test/unit/esm-named-exports.test.js:96]: ESM can be checked automatically by comparing itself to CJS, but checking CJS would require manually listing all CJS exports in this file. Could be considered but not part of this PR

File: test/unit/esm-named-exports.test.js (hunk 7 above)
Comment (ephys) [reply test/unit/esm-named-exports.test.js:10]: Done

File: src/instance-validator.js (hunk 6 above)
Comment (ephys) [reply src/instance-validator.js:5]: Done

File: src/decorators/legacy/validation.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+import type { OptionalParameterizedPropertyDecorator, RequiredParameterizedPropertyDecorator } from './decorator-utils.js';
+import {
+  createOptionallyParameterizedPropertyDecorator,
+  throwMustBeMethod,
+  throwMustBeModel,
+} from './decorator-utils.js';
+
+type ValidateKeys = Extract<keyof ColumnValidateOptions, string>;
+
+function createRequiredAttrValidateDecorator<Key extends ValidateKeys>(
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/decorators/legacy/validation.ts]: Done

File: test/unit/esm-named-exports.test.js (hunk 1 above)
Comment (evanrittenhouse) [reply test/unit/esm-named-exports.test.js:94]: Gotcha, resolving

File: src/decorators/legacy/decorator-utils.ts (hunk 4 above)
Comment (evanrittenhouse) [reply src/decorators/legacy/decorator-utils.ts:59]: Interesting! Thanks

**Review by evanrittenhouse — APPROVED** [review]

**Review by WikiRik — APPROVED** [review]

**Comment by ephys** [comment]
Thanks :)

**Pull request merged**